Complete revamp of binary output featuring new HDF5 interface

paulromano opened the pull request:
This pull request is a complete reworking of the binary output capabilities of OpenMC, represented previously by the output_interface, mpiio_interface, and hdf5_interface modules. output_interface provided a single unified API that would use either normal serial Fortran I/O, MPI-IO, or (parallel) HDF5 under the hood. For numerous reasons, I proposed in #429 that we move to using HDF5 only. This PR implements the necessary changes.

## Interface changes

For starters, output_interface and mpiio_interface are gone. There is now just a single hdf5_interface which has been heavily refactored. The API is now a flat API that is roughly inspired by Python's h5py API. It works as follows (fictitious example):
```Fortran
integer(HID_T) :: file_id, group_id

file_id = file_create("newfile.h5")
group_id = create_group(file_id, "tallies")
call write_dataset(group_id, "num_tallies", n_tallies)
call write_dataset(group_id, "num_meshes", n_meshes)
call close_group(group_id)
call write_dataset(file_id, "real_array", [1.0, 2.0, 8.0])
call write_dataset(file_id, "string_var", "This API is awesome!")
call file_close(file_id)
```
If you're familiar with the old output_interface API, you'll notice a few things. First, write_dataset is still generic, i.e., it will accept both scalars and arrays of various datatypes. Arrays no longer require explicitly specifying the shape. This took some hackery in hdf5_interface to get working correctly because the variable is eventually passed to `C_LOC` which requires that the variable is C interoperable. Although assumed-shape arrays are interoperable per F2008, gfortran 4.6 (which Travis still relies on) doesn't think so, so I have two sets of procedures, one with an `_explicit` suffix -- the first one gets passed an assumed-shape array and then passes it to a second routine which accepts an explicit-shape array. At some point in the future, we can combine the pairs of procedures.

The other thing you'll notice is that it is up to the user of the API to explicitly close groups whereas before this was done automatically. The downside of having it automatic was that the group was opened and closed each time a call to write_data was called. Now, the group is only opened once and its identifier is used in subsequent calls to write_dataset. This should result in less overhead as well as cleaner code (the state_point module is a good example).

Scalar variables are now actually written as scalars, not rank-1 arrays. This cleans up the Python backend a bit.

Another point to mention -- there are no more module variables in hdf5_interface that are reused, i.e., things like hdf5_err, dset, dspace, etc. Instead, each routine declares its own local variables. This is in an attempt to reduce the number of global variables we have, and is consistent with other recent changes in the code. It may also improve performance.

## CMake changes

Since the codebase is HDF5-only now, the HDF5 preprocessor macro is gone. One of the issues I wrestled with is having the ability to use both HDF5 and MPI simultaneously, without necessarily having to use PHDF5 (which may not be available to some users). CMakeLists.txt formerly relied on using the h5fc and h5pfc wrappers that are installed alongside HDF5/PHDF5. This would mean that if you want MPI and HDF5, you would have to use h5pfc, so the combination of serial HDF5 and MPI was not possible. Instead of that approach, I moved to using CMake's FindHDF5 module. However, there were a number of bugs in the existing FindHDF5 module. I've since fixed the bugs (now merged upstream -- see discussion [here](http://www.cmake.org/pipermail/cmake-developers/2015-August/026210.html)) but for our purposes, I've included the fixed version that we can use locally. That way, it shouldn't matter what version of CMake the user has. The most reliable way to pick a version of HDF5 is to set the HDF5_ROOT environment variable, e.g.
```sh
FC=mpif90 HDF5_ROOT=/opt/phdf5/1.8.15-gnu cmake ..
```

Another change I had to make in CMakeLists.txt -- there is a [bug in gfortran 4.6](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=49693) whereby module variables in common blocks can cause 'warning ___ defined but not used' warning messages upon compilation. This happens with HDF5 since there are unfortunately a number of common blocks. As a result, the output on Travis CI runs for debug configurations would be polluted with [tons of warnings](https://travis-ci.org/paulromano/openmc/jobs/78703124). To prevent those from being shown, the -Wall flag is not passed if gfortran 4.6 is being used.

There are a number of other improvements in CMakeLists.txt which are commented there, so I'll let the source speak for itself.

## Test changes

I've removed non HDF5 test configurations, as well as adding combinations of MPI and HDF5 that are not PHDF5.

## Remaining work

There are a few more things that need to be done before this PR is merged:

- [x] Update documentation
- [x] Test on various platforms/compilers
- [x] Decide what to do with openmc-tally-convergence script. It was hard coded based on the .binary format, so it is now obsolete. Either we need to update it to use the new openmc.statepoint or remove it altogether. I'm leaning towards the latter since I think the same thing could be accomplished with the new Python API pretty easily. Perhaps we ought to just have a Jupyter notebook demonstrating how to do it. (Thoughts @nelsonag?)

@bhermanmit If you have time, you would make a good reviewer for the new hdf5_interface, since you wrote the existing infrastructure.
@wbinventor You'll probably want to at least take a peek at the Python API changes.
@smharper I'd value your thoughts and comments.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -92,13 +91,8 @@ class StatePoint(object):
-        else:
-            self._f = open(filename, 'rb')
-            self._hdf5 = False
+        import h5py
```
Perhaps ``import h5py`` should be declared at the top of this file since HDF5 and ``h5py`` are no longer optional dependencies?

wbinventor replied:
On second thought, the Python API (minus ``openmc.statepoint``) could still be useful for input generation even if a user doesn't need to analyze HDF5 statepoint and summary files (*e.g.*, only the eigenvalue in the text output is needed). It's up to you whether you think ``h5py`` should be made a mandatory dependency by moving the import to the top of the file.

paulromano replied:
I had the same conflicting thoughts, and settled on leaving it where it is so that you don't absolutely need it just to construct input models.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -206,18 +197,15 @@ def _read_criticality(self):
-                 self._current_batch*self._gen_per_batch, path='entropy')
+            self._n_inactive = self._f['n_inactive'].value
+            self._gen_per_batch = self._f['gen_per_batch'].value
+            self._k_batch = self._f['k_generation'].value
```
Perhaps the ``_k_batch`` attribute should be changed to ``_k_generation`` to be consistent with the key in the statepoint file?

wbinventor commented on openmc/statepoint.py:
```diff
@@ -395,33 +358,30 @@ def _read_tallies(self):
 
                 if FILTER_TYPES[filter_type] == 'mesh':
-                    key = self._mesh_keys[self._mesh_ids.index(bins)]
+                    key = self._mesh_keys[list(self._mesh_ids).index(bins)]
```
Not a big deal, but perhaps take advantage of NumPy here:

```python
key = self._mesh_keys[self._mesh_ids == bins]
```

wbinventor commented on openmc/statepoint.py:
```diff
@@ -395,33 +358,30 @@ def _read_tallies(self):
-                 n_score_bins, path='{0}{1}/score_bins'.format(base, tally_key))]
-            n_user_scores = self._get_int(
-                 path='{0}{1}/n_user_score_bins'.format(base, tally_key))[0]
+            scores = [SCORE_TYPES[j] for j in self._f[
```
I think this is easier to comprehend in two lines:

```python
score_bins = self._f['{0}{1}/score_bins'.format(base, tally_key)].value
scores = [SCORE_TYPES[score] for score in score_bins]
```

wbinventor commented on scripts/openmc-track-to-vtk:
```diff
@@ -41,9 +42,8 @@ def main():
-            raise ValueError("Input file names must either end with '.h5' or"
-                             "'.binary'.")
+        if not fname.endswith('.h5'):
+            raise ValueError("Input file names must an HDF5 file.")
```
typo: "must an HD5F file" -> "must be an HDF5 file." I've seen some HDF5 files ending with ".hdf5" before, so might consider including that as an accepted option. And personally I think a message like "Input HDF5 files must have an ".h5" or ".hdf5"  extension" is more explicit and transparent than the current message.

wbinventor commented on openmc/particle_restart.py:
```diff
@@ -40,70 +40,31 @@ class Particle(object):
-        else:
-            self._f = open(filename, 'rb')
-            self._hdf5 = False
+        import h5py
```
See my comment below about importing ``h5py`` at global scale in this file.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -92,13 +91,8 @@ class StatePoint(object):
-            self._f = open(filename, 'rb')
-            self._hdf5 = False
+        import h5py
+        self._f = h5py.File(filename, 'r')
```
Perhaps include the error checking code for the file extension from the "openmc-track-to-vtk" script here.

paulromano replied:
I think for the Python API and scripts, there should be no expectation that the file extension is anything in particular. If the file is not an HDF5 file, the user will get an IOError telling them that the 'file signature is not found' which is the proper behavior as I see it. If you feel that error message is too cryptic, we could catch that exception and re-raise it with a more intuitive message.

wbinventor replied:
Sure, that makes sense to me.

wbinventor commented on openmc/particle_restart.py:
```diff
@@ -40,70 +40,31 @@ class Particle(object):
-            self._f = open(filename, 'rb')
-            self._hdf5 = False
+        import h5py
+        self._f = h5py.File(filename, 'r')
```
Perhaps include the error checking code for the file extension from the "openmc-track-to-vtk" script here.

wbinventor commented on scripts/openmc-track-to-vtk:
```diff
@@ -51,44 +51,20 @@ def main():
+        track = h5py.File(fname)
+        n_particles = track['n_particles'].value
+        n_coords = track['n_coords']
+        coords = []
```
This comment is unrelated to this PR, but I wonder if this block of code might be optimized by pre-allocating ``coords`` as a NumPy array outside of the loop rather than appending to a list on each iteration. Not a big deal at all, just couldn't help but express my affinity for NumPy.

paulromano replied:
Each item in coords is a 2D array that generally are not of uniform shape, so we couldn't really do a 3D array since it would be ragged.

wbinventor commented on scripts/openmc-track-to-vtk:
```diff
@@ -51,44 +51,20 @@ def main():
+        for i in range(n_particles):
+            coords.append(track['coordinates_' + str(i + 1)].value)
+            for j in range(n_coords[i]):
+                points.InsertNextPoint(coords[i][j,:])
```
Also a separate comment, but just noticed that ``InsertNextPoint`` doesn't conform to PEP8 :-)

paulromano replied:
We're lucky that VTK even works with Python 3 now! PEP8 conformance will probably take them another decade.

wbinventor replied:
Oh, I failed to realize that this routine is part of VTK and not a routine provided by this script.

wbinventor commented on src/hdf5_summary.F90:
```diff
@@ -27,48 +25,47 @@ module hdf5_summary
-    call hdf5_write_geometry()
-    call hdf5_write_materials()
-    call hdf5_write_nuclides()
+    call hdf5_write_geometry(file_id)
```
Just curious, since we are using all HDF5 now, any reason to explicitly (and redundantly) include "hdf5_" in these subroutine signatures?

wbinventor commented on src/hdf5_summary.F90:
```diff
@@ -97,784 +95,659 @@ end subroutine hdf5_write_header
       ! Write name for this cell
-      call su % write_data(c % name, "name", &
-           group="geometry/cells/cell " // trim(to_str(c % id)))
+      call write_dataset(cell_group, "name", c%name)
```
The Style Guide does not explicitly declare that whitespaces must be included around attributes (*e.g.*, ``c % name`` in place of ``c%name``) but this seems to be the way the majority of the code is written elsewhere.

wbinventor replied:
Or perhaps that is just how I wrote most of my code in "statepoint.F90" and "summary.F90", in which case stick to your preferred style.

paulromano replied:
These days I prefer no whitespaces when referencing derived type components, so I'm slowing moving in that direction. I may submit a pull request in the future changing all other modules to that style and adding it to the style guide, if others agree.

smharper replied:
I'm personally in favor of the whitespaces.  Something like `c%name` looks kind of cluttered...  I wish Fortran used a `.` like everybody else.

wbinventor replied:
I agree, I vote for the whitespaces, but of course it's just personal
preference.
On Sep 11, 2015 4:37 PM, "Sterling Harper" <notifications@github.com> wrote:

> In src/hdf5_summary.F90
> <https://github.com/mit-crpg/openmc/pull/447#discussion_r39314605>:
>
> >
> >        ! Write name for this cell
> > -      call su % write_data(c % name, "name", &
> > -           group="geometry/cells/cell " // trim(to_str(c % id)))
> > +      call write_dataset(cell_group, "name", c%name)
>
> I'm personally in favor of the whitespaces. Something like c%name looks
> kind of cluttered... I wish Fortran used a . like everybody else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/447/files#r39314605>.
>

wbinventor commented:
@paulromano from my brief overview of this PR, it looks like a great improvement to the codebase. I'm very much in support of this move to HDF5, though there may be some growing pains for users who have  up-to-now relied on the standard ASCII output files. 

One nice aspect of this transition is that it will permit indexing of HDF5 files with memory footprints that approach or exceed that available on a single node. For example, each ``Tally`` in the Python API might read its data on-the-fly from a statepoint file only when accessed through the property decorator (*e.g.*, ``Tally.sum``, ``Tally.mean``, etc.). The ``StatePoint.read_results()`` routine could be eliminated. Instead, a reference (or filename string) to the HDF5 statepoint file might be stored in each ``Tally`` object so that they can independently load their data if and when necessary. I'm not sure if you want to implement this on-the-fly tally data allocation for this PR, but it seems like one enhancement that is made possible by the changes you've made here.

bhermanmit commented:
Looks good to me! I would suggest just adding the other two build configurations to Travis: omp-hdf5-debug, mpi-hdf5-debug -- unless you think they are not needed.

nelsonag commented:
@paulromano  I agree with removing the tally convergence script; the API could accomplish the same thing quite easily and any user who is curious about some convergence traits probably has a special use case in mind in which this general script wouldn't satisfy the needs of.

smharper commented:
Comments concerning hdf5_interface.F90:

The header is kind of nonstandard since it's missing the `===` bars and the module name.

Line 141 has `if (trim(mode) == 'w') open_mode = H5F_ACC_RDWR_F`; it might be safer to change it to `if (trim(to_lower(mode)) == 'w') open_mode = H5F_ACC_RDWR_F`.

`open_group` should have something like `if (.not. exists) call fatal_error()`

`create_group` should similarly `if (exists) call fatal_error()`

`close_group` should probably also check for the existence of the group.

paulromano commented:
@smharper I agree with all your points except for the one about line 141. For example, in Python `open('newfile', 'w')` works but `open('newfile', 'W')` does not. Relooking at that line though, I realized that calling trim() is not even necessary because string comparisons treat unequal-length variables such that the shorter one is blank-padded at the end, so now the comparison is just `mode == 'w'`.

@nelsonag I've removed the tally convergence script.

smharper commented:
Okay, your point about `to_lower` on line 141 makes sense.  It looks like you missed `close_group` with the existence checking.  Other than `close_group`, I'm happy with the Fortran chunk of this PR.

paulromano commented:
@smharper I added a check on hdf5_err in close_group. It would require more HDF5 API calls to get the group name and check if it's valid and open, so instead I take the approach of simply trying to close it and showing an error if it doesn't complete successfully.

paulromano commented:
The documentation is now up-to-date. I've changed installation instructions to reflect the fact that HDF5 is a requirement and give some pointers on configuring it properly. I've also revised the binary file format documentation so that it shows the file structure of each HDF5 output file and moved those from the developer's guide to the user's guide since not only developers will want to know what's in their output files!

smharper commented:
:+1:

paulromano commented:
I've started doing some testing on various platforms to help preempt any potential problems. One that was particularly difficult was compiling on the Xeon Phi. To get HDF5 compiled for the Xeon Phi requires a number of hacks. I set up a [script](https://github.com/paulromano/install-scripts/blob/master/install-hdf5-mic) that is based on two previous accounts ([1](http://linux4hippos.blogspot.co.id/2013/05/cross-compile-hdf5-for-intel-xeon-phi.html) and [2](https://github.com/gahansen/Albany/wiki/Westley-mic)). It works, but is not terribly satisfying. I may take up the issue with the HDF5 group directly to try to ease the process for future releases.

At the moment, Blue Gene/Q at ANL (Mira/Cetus) doesn't have a build of HDF5 with --enable-fortran2003, which we need because of heavy use of `c_loc` and `type(c_ptr)`. ALCF support is working on getting a new build there so that should not be an issue either.

cjosey commented:
The Xeon Phi HDF5 stuff should be a bit easier than that first account makes it out to be.  The issue is that the buildsystem creates executables that probe the architecture it's run on and outputs source code specifically targeting said architecture.  A more "proper" modification simply requires the 3 or so commands it fails to run to be ssh executed as opposed to normally executed.  Doubly compiling and using the Xeon versions of the output isn't needed, though it is easier to automate.

Of course, this should cease mattering when the 3rd gen Phi comes out, as all I'm hearing is that it'll be ABI/instruction compatible with Skylake-E.

paulromano commented:
@cjosey Very good points. If KNL does really use the same instruction set (I've heard the same), that will indeed make this all a moot point.

paulromano commented:
I'm satisfied with this branch now and believe it is ready for merging. I made a few more small commits, including one that ensures OpenMC will compile with the PGI Fortran compiler (apparently if you have a local variable name that is the same as a module name, it doesn't like it). I've tested the branch with gfortran 4.6/4.9, Intel Fortran 15.0.2, PGI Fortran 15.7, and IBM XLF on BG/Q. I've also run on x86_64, ppc64, and k1om (MIC) architectures.

wbinventor commented:
@smharper you may be best prepared to merge this one in since I only had time to look over the changes to the Python API.

smharper commented:
Merged.  Good work, team!

wbinventor commented:
I wanted to comment on a few issues that some users may face as they adjust to the changes made in this PR. First, make sure to take heed to @paulromano's comments on the new method to build the code with HDF5 (and MPI) at the top of this thread. This is particularly important since the online documentation - in particular, the build instructions - no longer apply to the current develop branch. If you pull the latest develop branch, you **must** reconfigure CMake in order for the build to work properly, *e.g.*:

```
rm -rf build
mkdir build
FC=mpif90 HDF5_ROOT=/opt/phdf5/1.8.14-gnu cmake ..
```

In addition, note that the default version of ``sphinx`` installed by ``apt-get`` on Ubuntu 14.10 will no longer build the code documentation which relies on the newish ``napoleon`` submodule. sphinx >= 1.3 must be installed, which can be easily installed with ``pip`` as discussed in the [Developers Guide](https://mit-crpg.github.io/openmc/devguide/docbuild.html).

smharper commented:
Hit a little snag with compilation.  If I set `FC = h5pfc`, then our CMake doesn't set `MPI_ENABLED` and `-DMPI`.  Is this an error?

smharper commented:
Should be noted that this results in an exception, "Parallel HDF5 must be used with MPI."

wbinventor commented:
Yeah, I hit that snag as well. I instead set FC=mpif90 and let cmake find
my HDF5 install on its own (which it did, albeit it found the serial
version without HDF5_ROOT) and if compiled for parallel runs.
On Sep 14, 2015 9:11 PM, "Sterling Harper" <notifications@github.com> wrote:

> Hit a little snag with compilation. If I set FC = h5pfc, then our CMake
> doesn't set MPI_ENABLED and -DMPI. Is this an error?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/447#issuecomment-140245324>.
>

paulromano commented:
@smharper @wbinventor The new FindHDF5 module will preferentially find the serial version. If you have both a serial and parallel version on your PATH and you want the parallel one to be preferred, you can pass -DHDF5_PREFER_PARALLEL to cmake. Or you can just set HDF5_ROOT which has the same effect.

The reason I moved away from using the HDF5 wrappers directly is that it doesn't allow for using serial HDF5 + MPI, which may be necessary on some systems.

smharper commented:
Okay, I am satisfied.
